layers/tls_handshake.go

```diff
@@ -7,12 +7,119 @@
+	TLSHandshakeCertificateRequest  = 13
+	TLSHandshakeServerHelloDone     = 14
+	TLSHandshakeCertificateVerify   = 15
+	TLSHandshakeClientKeyExchange   = 16
+	TLSHandshakeFinished            = 20
+)
+
+var handShakeTypeMap = map[uint8]string{
+	TLSHandshakeHelloRequest:        "Hello Request",
+	TLSHandshakeClientHello:         "Client Hello",
+	TLSHandshakeServerHello:         "Server Hello",
+	TLSHandsharkHelloVerirfyRequest: "Hello Verify Request",
+	TLSHandshakeCertificate:         "Certificate",
+	TLSHandshakeServerKeyExchange:   "Server Key Exchange",
+	TLSHandshakeCertificateRequest:  "Certificate Request",
+	TLSHandshakeServerHelloDone:     "Server Hello Done",
+	TLSHandshakeCertificateVerify:   "Certificate Verify",
+	TLSHandshakeClientKeyExchange:   "Client Key Exchange",
+	TLSHandshakeFinished:            "Finished",
+}
+
+type TLSHandshakeRecordClientHello struct {
+	HandshakeType            uint8
+	Length                   uint32
+	ProtocolVersion          TLSVersion
+	Random                   []uint8
+	SessionIDLength          uint8
+	SessionID                []uint8
+	CipherSuitsLength        uint16
+	CipherSuits              []uint8
+	CompressionMethodsLength uint8
+	CompressionMethods       []uint8
+	ExtensionsLength         uint16
+	Extensions               []uint8
+}
+
+type TLSHandshakeRecordClientKeyChange struct {
+}
+
 // TLSHandshakeRecord defines the structure of a Handshare Record
 type TLSHandshakeRecord struct {
 	TLSRecordHeader
+	ClientHello     TLSHandshakeRecordClientHello
+	ClientKeyChange TLSHandshakeRecordClientKeyChange
+}
+
+func (t *TLSHandshakeRecordClientHello) decodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	t.HandshakeType = data[0]
+	d := make([]byte, 4)
+	for k, v := range data[1:4] {
+		d[k+1] = v
+	}
+	t.Length = binary.BigEndian.Uint32(d)
+	t.ProtocolVersion = TLSVersion(binary.BigEndian.Uint16(data[4:6]))
+	t.Random = data[6:38]
+	t.SessionIDLength = data[38]
+	t.SessionID = data[39 : 39+t.SessionIDLength]
+	t.CipherSuitsLength = binary.BigEndian.Uint16(data[39+t.SessionIDLength : 39+t.SessionIDLength+2])
+	t.CipherSuits = data[39+t.SessionIDLength+2 : (39 + uint16(t.SessionIDLength) + 2 + t.CipherSuitsLength)]
+	t.CompressionMethodsLength = data[(39 + uint16(t.SessionIDLength) + 2 + t.CipherSuitsLength)]
+	t.CompressionMethods = data[(39+uint16(t.SessionIDLength)+2+t.CipherSuitsLength)+1 : (39+uint16(t.SessionIDLength)+2+t.CipherSuitsLength)+1+uint16(t.CompressionMethodsLength)]
+	t.ExtensionsLength = binary.BigEndian.Uint16(data[(39+uint16(t.SessionIDLength)+2+t.CipherSuitsLength)+1+uint16(t.CompressionMethodsLength) : (39+uint16(t.SessionIDLength)+2+t.CipherSuitsLength)+1+uint16(t.CompressionMethodsLength)+2])
+	t.Extensions = data[((39 + uint16(t.SessionIDLength) + 2 + t.CipherSuitsLength) + 1 + uint16(t.CompressionMethodsLength) + 2) : ((39+uint16(t.SessionIDLength)+2+t.CipherSuitsLength)+1+uint16(t.CompressionMethodsLength)+2)+t.ExtensionsLength]
```

Comment:
in here, if the order of these changes in the Client Hello packet, what would be the behavior of the code. Let's assume CipherSuits and SessionID are sent in reverse.. hope that explains it

Reply:
CipherSuits and SessionID are not TLV (Type/Length/Value)structures. Then can not be sent in reverse. 
Only Extension has TLV structure.